### PR TITLE
Cherry-pick #8363 to r293

### PR DESCRIPTION
**Pull request by pstibrany**
#### What this PR does

This PR cherry-picks PR #8363 to `r293` branch.

Cherry-pick was done manually, since there are incompatible changes on `main`.

#### Checklist

- [ ] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: vendor/github.com/grafana/dskit/cache/memcached_client.go [hunk 1]
```diff
@@ -315,22 +315,27 @@ func (c *MemcachedClient) Name() string {
...
-	c.setAsync(key, value, ttl, func(key string, buf []byte, ttl time.Duration) error {
-		return c.client.Set(&memcache.Item{
-			Key:        key,
-			Value:      buf,
-			Expiration: int32(ttl.Seconds()),
-		})
+	c.setAsync(key, value, ttl, c.setSingleItem)
+}
+
+func (c *MemcachedClient) setSingleItem(key string, value []byte, ttl time.Duration) error {
```
Comment (aknuds1) [inline vendor/github.com/grafana/dskit/cache/memcached_client.go:325]: I see this brings in unrelated dskit changes, should we possibly use a dskit revision based on the one used in r293??

Comment (pstibrany) [reply vendor/github.com/grafana/dskit/cache/memcached_client.go]: I've switched dskit to version from `r293` + our fix.

**Review by aknuds1 — APPROVED** [review]
